### PR TITLE
Replace most assert_throws uses with assert_throws_dom and assert_throws_js

### PR DESCRIPTION
**Pull request by bzbarsky**
The individual commit messages explain how each commit was generated.

### CODE REVIEW

**Comment by bzbarsky** [comment]
@jgraham @foolip

**Comment by bzbarsky** [comment]
@stephenmcgruer

This has a tendency to bitrot quickly, so speedy reviews would be much appreciated.  And yes, I know it's a pain to review because of the size.  :(

Fwiw, with this applied there are 208 calls to `assert_throws` left.

**Comment by stephenmcgruer** [comment]
Reviewing now.

File: xhr/resources/send-after-setting-document-domain-window-helper.js
Comment (atotic) [inline xhr/resources/send-after-setting-document-domain-window-helper.js]: Might be nice to put unexpected error name in the error message:

    throw `Exception '${e.constructor.name || "unknown"}' that was not a DOMException was thrown`;

Comment (bzbarsky) [reply xhr/resources/send-after-setting-document-domain-window-helper.js]: Sure thing, will do.

**Review by atotic — APPROVED** [review]
The only gotcha with these changes is finding the right constructor to pass to assert_throws_js when there are multiple documents. Might be nice to document how to do this, I can see being annoyed when my assert is failing when I am clearly passing in the right kind of an Exception.

**Comment by bzbarsky** [comment]
Yeah, the "right constructor" thing is a PITA.  That's one reason the idlharness changes are not part of the mass-change...

That said, part of the point is being able to test that the exceptions come from the global the spec says to use, so we really do want to have that tested.  I'll try to think about whether we can make the error messages more informative...

**Comment by bzbarsky** [comment]
Hmm.  The azure pipelines tests seem to be failing due to pulling one of the pre-update (hence now gone) changesets for this PR...  Not sure how to retrigger them on the current set of changesets.

**Comment by atotic** [comment]
> Hmm. The azure pipelines tests seem to be failing due to pulling one of the pre-update (hence now gone) changesets for this PR... Not sure how to retrigger them on the current set of changesets.

That sounds like a weird edge case. Manual merge instead?

**Comment by bzbarsky** [comment]
I can certainly do that.  This doesn't need more reviews from anyone else?

**Comment by stephenmcgruer** [comment]
You're going to fail the TC Firefox stability check anyway, so this PR will (eventually) have to be admin-merged. However, my review is still ongoing :P

**Comment by atotic** [comment]
> I can certainly do that. This doesn't need more reviews from anyone else?

I am a junior here, so I do not know. 
My understanding is that my approval gives you authority to perform the merge, github is ok with it.
I do not know whether other reviewers would like to have a say before you do, you requested reviews from everyone,

**Comment by stephenmcgruer** [comment]
> you requested reviews from everyone,

An unfortunate side-effect of the way that wpt-pr-bot and GitHub work. GitHub has no notion of 'cc' versus 'reviewers', so wpt-pr-bot has to add everyone who has asked to be notified of changes to the reviewers list. It then randomly chooses one person to be the 'assignee', and hopes that that person will review the code.

In this case, the change touches half the directories in WPT, so wpt-pr-bot went a little crazy :D

**Comment by bzbarsky** [comment]
It's not 100% obvious to me that stability checks will fail, through I would be pretty surprised if they don't, yeah...  Will wait on @stephenmcgruer's review, then see what I can do to get this merged.  ;)

**Comment by stephenmcgruer** [comment]
> It's not 100% obvious to me that stability checks will fail

The change will affect too many tests, and at the very least Firefox stability check will not complete in time (probably Chrome will also timeout); they run every affected test 10 times minimum :D

> then see what I can do to get this merged.

I can take care of this; magic admin powers :)

**Comment by bzbarsky** [comment]
> I can take care of this; magic admin powers :)

Lovely, thank you!

**Comment by stephenmcgruer** [comment]
Closing/re-opening to retrigger the CI. There's some concerning failures from TC that I want to see if they reoccur, and hopefully this will fix Safari too.

**Comment by bzbarsky** [comment]
@stephenmcgruer Did you create a separte PR for https://github.com/web-platform-tests/wpt/pull/21350/commits/8f8f74f86c08cb97f1e3f2c9aaa803a0ff53699d too?  I'm not finding it so far.

**Comment by stephenmcgruer** [comment]
@bzbarsky whoops, overlooked it (didn't see single quotes vs double). Will do so now.

**Comment by bzbarsky** [comment]
Awesome, thanks!

**Comment by bzbarsky** [comment]
@stephenmcgruer landed the pieces of this in separate PRs, so this is all done now.

I'll do separate, smaller, PRs for the remaining bits.

**Comment by alvestrand** [comment]
Unfortunately when this was imported into Chrome's test framework, the tests (specific example: webrtc/RTCConfiguration-iceServes.html) broke. It seems that assert_throws_js and assert_throws_dom aren't defined in the version of testharness.js that Chrome runs.

**Comment by alvestrand** [comment]
Turns out some Chrome setup gets the wrong copy of testharness.js. Chrome problem, I guess.

**Review by Phum7895 — APPROVED** [review]

**Pull request closed without merging**
